### ISSUE TITLE
Quest: How to import custom functions under WASI

### ISSUE THREAD

**Issue by mrpre** [WAVM/WAVM#180]
WASI says it can import other host functions except for `wasi-core`
![image](https://user-images.githubusercontent.com/5129205/61858160-1f31f180-aef8-11e9-94f6-9168fa0b928d.png)
Is there any interface to import host functions ?

**Comment by AndrewScheidecker** [comment]
> 1. Is there any interface to import host functions and sample to show the usage?

WAVM's WASI module doesn't currently expose a way for C++ code to export functions to the WASM code. I'll try to figure out a way to expose that.

> 2. I find c-api has been supported. Will c-api support both wasi/emscripten/ imports and custom imports?

The C API does not currently support using WAVM's implementations of the Emscripten/WASI host functions. For now, the C API is pretty close to https://github.com/WebAssembly/wasm-c-api, but I'd like to expose more of WAVM's functionality through it.

The C API does support linking C functions to WASM module imports using `wasm_func_new` and `wasm_func_as_extern` to create a `wasm_extern_t` for the C function, and then passing it to `wasm_instance_new` in the imports array.

**Comment by mrpre** [comment]
@AndrewScheidecker How about Emscripten module ?  

1.   I'm using this way to import `my_log` so that the Wasm code can call it. Am i right? 
```
WAVM_DECLARE_INTRINSIC_MODULE(env);
WAVM_DEFINE_INTRINSIC_FUNCTION(env, "_my_log", void, my_log, I32 wstr)
{
}

```  

2. How to set personal userdata while using Emscripten module ?
Seems that the Emscripten has already set it's own userdata by `setUserData` which I can't replace it.  

https://github.com/WAVM/WAVM/blob/2c4fbd0e82f8b162862a7cd6a9cb781f38e71c50/Lib/Emscripten/Emscripten.cpp#L1140-L1145   

I'm try to use my personal context now like this:  
```
WAVM_DEFINE_INTRINSIC_FUNCTION(env, "_my_log", void, my_log, I32 wstr)
{
    //Seems that the userdata I get is the Emscripten instance, Is there any possibility to set my own ctx and get it in callback ?
     personal_ctx_t = (personal_ctx_t)*getUserdata(...);
}
```

**Comment by AndrewScheidecker** [comment]
> @AndrewScheidecker How about Emscripten module ?
> 
> 1. I'm using this way to import `my_log` so that the Wasm code can call it. Am i right?
> 
> 
> ```
> WAVM_DECLARE_INTRINSIC_MODULE(env);
> WAVM_DEFINE_INTRINSIC_FUNCTION(env, "_my_log", void, my_log, I32 wstr)
> {
> }
> ```

That's correct, but not sufficient. You also need to instantiate your intrinsic module and link its exports to the WASM module's imports. e.g.:

https://github.com/WAVM/WAVM/blob/d542d7d5ce72a22c8cd4abfa1213cbc7f4b88a15/Lib/Emscripten/Emscripten.cpp#L1125

https://github.com/WAVM/WAVM/blob/d542d7d5ce72a22c8cd4abfa1213cbc7f4b88a15/Programs/wavm-run/wavm-run.cpp#L155

https://github.com/WAVM/WAVM/blob/d542d7d5ce72a22c8cd4abfa1213cbc7f4b88a15/Programs/wavm-run/wavm-run.cpp#L171

Sorry to say that the simplest way to do this is to just modify WAVM's Emscripten module to add the functions you want to the "env" module it declares.

It should be possible to change WAVM to make `Emscripten::instantiate` take a set of intrinsic modules to merge into the "env" module. [This `Intrinsics::instantiate` call](https://github.com/WAVM/WAVM/blob/d542d7d5ce72a22c8cd4abfa1213cbc7f4b88a15/Lib/Emscripten/Emscripten.cpp#L1125) can do the merge if multiple intrinsic modules are passed to it. However, that would not solve your second problem.

> 2. How to set personal userdata while using Emscripten module ?
>        Seems that the Emscripten has already set it's own userdata by `setUserData` which I can't replace it.

There isn't a great way to do this right now. There are a few possibilities:

1. Like the "simple" solution to your first problem, just modify WAVM's Emscripten module and put the necessary information in its context.
2. Add another userData pointer to `Emscripten::Instance`.
3. Add intrinsic closures: i.e. functions that take a context argument, and must be bound to some value for it when they are instantiated.
